### PR TITLE
fix mypy in downloader.py

### PR DESCRIPTION
**Pull request by jochman**
RDY
fixes:
```python
demisto_sdk/commands/download/downloader.py:73: error: Need type annotation for 'all_custom_content_objects' (hint: "all_custom_content_objects: List[<type>] = ...")
demisto_sdk/commands/download/downloader.py:74: error: Need type annotation for 'files_not_downloaded' (hint: "files_not_downloaded: List[<type>] = ...")
demisto_sdk/commands/download/downloader.py:75: error: Need type annotation for 'custom_content' (hint: "custom_content: List[<type>] = ...")
demisto_sdk/commands/download/downloader.py:76: error: Need type annotation for 'pack_content'
demisto_sdk/commands/download/downloader.py:133: error: Incompatible types in assignment (expression has type "str", variable has type "None")
demisto_sdk/commands/download/downloader.py:137: error: No overload variant of "join" matches argument types "None", "str"
demisto_sdk/commands/download/downloader.py:137: note: Possible overload variants:
demisto_sdk/commands/download/downloader.py:137: note:     def join(a: Union[str, _PathLike[str]], *paths: Union[str, _PathLike[str]]) -> str
demisto_sdk/commands/download/downloader.py:137: note:     def join(a: Union[bytes, _PathLike[bytes]], *paths: Union[bytes, _PathLike[bytes]]) -> bytes
demisto_sdk/commands/download/downloader.py:139: error: Item "None" of "Optional[IO[bytes]]" has no attribute "read"
demisto_sdk/commands/download/downloader.py:172: error: Attribute 'all_custom_content_objects' already defined on line 73
demisto_sdk/commands/download/downloader.py:219: error: Name 'entity_instances_paths' already defined on line 217
demisto_sdk/commands/download/downloader.py:346: error: Argument "mypytype" to "find_type" has incompatible type "Optional[str]"; expected "str"
demisto_sdk/commands/download/downloader.py:347: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")
demisto_sdk/commands/download/downloader.py:514: error: Name 'dir_output_path' already defined on line 511
```

### CODE REVIEW

File: demisto_sdk/commands/download/downloader.py [hunk 1]
```diff
@@ -136,7 +137,13 @@ def fetch_custom_content(self) -> bool:
...
                 file_path: str = os.path.join(self.custom_content_temp_dir, file_name)
                 with open(file_path, 'w') as file:
-                    file.write(tar.extractfile(member).read().decode('utf-8'))
+                    extracted_file = tar.extractfile(member)
+                    # File might empty
+                    if extracted_file:
+                        string_to_write = extracted_file.read().decode('utf-8')
+                        file.write(string_to_write)
+                    else:
+                        raise FileNotFoundError(f'Could not extract files from tar file: {file_path}')
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:146]: i think we need to print error here and return false
@guyfreund please review

File: demisto_sdk/commands/download/downloader.py [hunk 2]
```diff
@@ -69,11 +70,11 @@ def __init__(self, output: str, input: str, force: bool = False, insecure: bool
         self.all_custom_content = all_custom_content
         self.run_format = run_format
         self.client = None
-        self.custom_content_temp_dir = None
-        self.all_custom_content_objects = list()
-        self.files_not_downloaded = list()
-        self.custom_content = list()
-        self.pack_content = {entity: list() for entity in CONTENT_ENTITIES_DIRS}
+        self.custom_content_temp_dir: Optional[str] = None
+        self.all_custom_content_objects: List[dict] = list()
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:74]: why are you not using `List` and `Dict` in the nested types?

File: demisto_sdk/commands/download/downloader.py [hunk 3]
```diff
@@ -216,7 +225,7 @@ def build_pack_content(self) -> None:
                 # If entity is of type integration/script it will have dirs, otherwise files
                 entity_instances_paths: list = get_child_directories(content_entity_path)
             else:
-                entity_instances_paths: list = get_child_files(content_entity_path)
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:219]: why did you change this? whats returned from the func?

File: demisto_sdk/commands/download/downloader.py [hunk 4]
```diff
@@ -300,7 +309,7 @@ def build_custom_content(self) -> None:
         Build a data structure called pack content that holds basic data for each content entity instances downloaded from Demisto.
         For example check out the CUSTOM_CONTENT variable in downloader_test.py
         """
-        custom_content_objects: list = self.all_custom_content_objects if self.all_custom_content_objects else\
+        custom_content_objects = self.all_custom_content_objects if self.all_custom_content_objects else\
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:312]: why did you change this? whats returned from the func?

File: demisto_sdk/commands/download/downloader.py [hunk 5]
```diff
@@ -344,7 +353,9 @@ def build_custom_content_object(file_path: str) -> dict:
         """
         file_data, file_ending = get_dict_from_file(file_path)  # For example: yml, for integration files
         file_type: str = find_type(_dict=file_data, file_type=file_ending)  # For example: integration
-        file_entity: str = ENTITY_TYPE_TO_DIR.get(file_type)  # For example: Integrations
+        file_entity = ENTITY_TYPE_TO_DIR.get(file_type)  # For example: Integrations
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:356]: why did you change this? whats returned from the func?

File: demisto_sdk/commands/download/downloader.py [hunk 6]
```diff
@@ -597,7 +608,7 @@ def update_data(file_path_to_write: str, file_path_to_read: str, file_ending: st
         :return: None
         """
         ryaml = YAML()
-        ryaml.preserve_quotes = True
+        ryaml.preserve_quotes = True  # type: ignore
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:611]: why?

File: demisto_sdk/commands/download/downloader.py [hunk 7]
```diff
@@ -511,7 +522,7 @@ def merge_and_extract_new_file(self, custom_content_object: dict) -> None:
         dir_output_path: str = os.path.join(self.output_pack_path, file_entity)
         # dir name should be the same as file name without separators mentioned in constants.py
         dir_name: str = self.create_dir_name(file_name)
-        dir_output_path: str = os.path.join(dir_output_path, dir_name)
+        dir_output_path = os.path.join(dir_output_path, dir_name)
```
Comment (Itay4) [inline demisto_sdk/commands/download/downloader.py:525]: why did you change this? whats returned from the func?

File: demisto_sdk/commands/download/downloader.py (hunk 1 above)
Comment (guyfreund) [reply demisto_sdk/commands/download/downloader.py:146]: There's no need for this error handling since the file doesn't exist, we are creating it here.

File: demisto_sdk/commands/download/downloader.py [hunk 8]
```diff
@@ -169,8 +176,10 @@ def handle_list_files_flag(self) -> bool:
         :return: True if list-files flag is on and listing available files process succeeded, False otherwise
         """
         if self.list_files:
-            self.all_custom_content_objects: list = self.get_custom_content_objects()
-            list_files: list = [[cco['name'], cco['entity'][:-1]] for cco in self.all_custom_content_objects]
+            self.all_custom_content_objects = self.get_custom_content_objects()
+            list_files = [
```
Comment (guyfreund) [inline demisto_sdk/commands/download/downloader.py:180]: Let's annotate here too

File: demisto_sdk/commands/download/downloader.py (hunk 3 above)
Comment (guyfreund) [reply demisto_sdk/commands/download/downloader.py:219]: List is returned and we should annotate

File: demisto_sdk/commands/download/downloader.py (hunk 4 above)
Comment (guyfreund) [reply demisto_sdk/commands/download/downloader.py:312]: both values are lists, why removed?

File: demisto_sdk/commands/download/downloader.py (hunk 5 above)
Comment (guyfreund) [reply demisto_sdk/commands/download/downloader.py:356]: ENTITY_TYPE_TO_DIR dict contains only strings, what is wrong here?

File: demisto_sdk/commands/download/downloader.py [hunk 9]
```diff
@@ -344,7 +353,9 @@ def build_custom_content_object(file_path: str) -> dict:
         """
         file_data, file_ending = get_dict_from_file(file_path)  # For example: yml, for integration files
         file_type: str = find_type(_dict=file_data, file_type=file_ending)  # For example: integration
-        file_entity: str = ENTITY_TYPE_TO_DIR.get(file_type)  # For example: Integrations
+        file_entity = ENTITY_TYPE_TO_DIR.get(file_type)  # For example: Integrations
+        if not file_entity:
+            raise KeyError(f'{file_path}: Could not find entity type in file.')
```
Comment (guyfreund) [inline demisto_sdk/commands/download/downloader.py:358]: No need for this since we cover all content entities

File: demisto_sdk/commands/download/downloader.py (hunk 1 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:146]: it returns false in the catch.
MYPY is yelling on this

File: demisto_sdk/commands/download/downloader.py (hunk 3 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:219]: we should not double annotate.
image

File: demisto_sdk/commands/download/downloader.py (hunk 4 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:312]: because you don't need this annotation, mypy knows.

File: demisto_sdk/commands/download/downloader.py (hunk 5 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:356]: it can return None.

File: demisto_sdk/commands/download/downloader.py (hunk 6 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:611]: because of ryaml annotations is faulty. (`preserve_quotes` Is None and not bool)

File: demisto_sdk/commands/download/downloader.py (hunk 7 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:525]: Double annotation.
image

File: demisto_sdk/commands/download/downloader.py (hunk 8 above)
Comment (jochman) [reply demisto_sdk/commands/download/downloader.py:180]: Shouldn't do so, it's clearly a list.
